### PR TITLE
[Issue-4266][pulsar-flink] Flink Pulsar Producer should reused across threads

### PR DESCRIPTION
**Pull request by Xpray**
Master Issue: #4266

### Motivation

* Reuse Pulsar Producer for Apache Flink Connector.

### Modifications

* Make Producer static

### Verifying this change

- [x] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency):  no
  - The public API:  no
  - The schema:  no
  - The default values of configurations: no
  - The wire protocol:  no
  - The rest endpoints:  no
  - The admin cli options:  no
  - Anything that affects deployment:  no

### Documentation

  - Does this pull request introduce a new feature?  no
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

**Comment by jiazhai** [comment]
run java8 tests
run integration tests

File: pulsar-flink/src/main/java/org/apache/flink/batch/connectors/pulsar/BasePulsarOutputFormat.java [hunk 1]
```diff
@@ -85,7 +90,8 @@ public void configure(Configuration configuration) {
 
     @Override
     public void open(int taskNumber, int numTasks) throws IOException {
-        this.producer = getProducerInstance();
+        producer = getProducerInstance();
+        producerRefCnt.incrementAndGet();
```
Comment (jiazhai) [inline pulsar-flink/src/main/java/org/apache/flink/batch/connectors/pulsar/BasePulsarOutputFormat.java]: How about move this ref change into the `synchronized` block of `getProducerInstance`?

File: pulsar-flink/src/main/java/org/apache/flink/streaming/connectors/pulsar/FlinkPulsarProducer.java [hunk 2]
```diff
@@ -200,7 +211,8 @@ public void setFlushOnCheckpoint(boolean flush) {
      */
     @Override
     public void open(Configuration parameters) throws Exception {
-        this.producer = createProducer();
+        createProducer();
+        producerRefCnt.incrementAndGet();
```
Comment (jiazhai) [inline pulsar-flink/src/main/java/org/apache/flink/streaming/connectors/pulsar/FlinkPulsarProducer.java]: Maybe also here? move ref change into create producer?

**Comment by Xpray** [comment]
@jiazhai thanks for you kind review, I've updated the PR, since the refCnt is Atomic I think it's better be used outside the `synchronized` block, what do you think? Best

**Comment by Xpray** [comment]
For some unknown reasons it did not pass Integration Tests and Java8 Unit Tests, but I checked the details found that it fails due to other tests.

**Comment by codelipenghui** [comment]
@Xpray Sorry, I didn't understand what the main purpose of this improvement is. Can you add some detailed information? This will helps us to understand.

**Comment by Xpray** [comment]
@jiazhai , It's my fault, here should be a mapping from topic to producer, I'll take care of it.

**Comment by Xpray** [comment]
@codelipenghui  I intend to make producer shared across Flink tasks, 
A TaskManager of Flink may consists of several tasks, each task will have a Producer for now, apply this changes can reduce the connections to Pulsar, but here is a fatal bug in my PR, I'll do some changes. Best.

**Comment by codelipenghui** [comment]
@Xpray Is your goal to share the producer instance between tasks which is in the same job of the same TaskManager? Can you explain in more detail how flink can share instance between tasks? I only know that the same job id in the flink TaskManager will share classloader.

If i have some wrong places, please point out, thanks.

**Comment by Xpray** [comment]
@codelipenghui. Thanks for your review. Task is an execution of a sub graph of the whole job,  and a TaskManager(TM) is a container, which can support running multiple tasks(maybe from different Jobs) together, all tasks of a same Job in the same TM will share the same ClassLoader, each of the tasks will have its own main thread, so any static field can be shared between these threads.

**Comment by codelipenghui** [comment]
@Xpray Thanks, this make sense to me. Can we let the user decide share the producer, user can define a static producer, and we just allow user use their own producer to create a FlinkPulsarProducer instance. This is more transparent to the user and considering that the producer has a lot of configuration options, maybe the same name as the topic, but the other configurations are different.

**Comment by jiazhai** [comment]
run java8 tests

**Comment by Xpray** [comment]
@codelipenghui It make sense. I'm trying to figure out a simple way for user to specify if or nor they want Producer to be shared.

**Comment by sijie** [comment]
@Xpray any updates on this pull request? or do we consider moving this to 2.5.0?

**Comment by Xpray** [comment]
> @Xpray any updates on this pull request? or do we consider moving this to 2.5.0?

Hi @sijie , I think I need more time for this issue. It's better to move to next version. Best

**Comment by aahmed-se** [comment]
@Xpray can you rebase this pr.

**Comment by sijie** [comment]
this has been fixed by #5900

**Pull request closed without merging**